pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -187,3 +237,140 @@ func getPacketInfo(pktIn *ofctrl.PacketIn, ob *logInfo) error {
 	}
 	return nil
 }
+
+// rejectRequest sends reject response to the requesting client, based on the
+// packet-in message.
+func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
+
+	var (
+		srcMAC     net.HardwareAddr
+		dstMAC     net.HardwareAddr
+		srcIP      string
+		dstIP      string
+		IPProtocol uint8
+		TCPSrcPort uint16
+		TCPDstPort uint16
+		TCPFlags   uint8
+		TCPSeqNum  uint32
+		TCPAckNum  uint32
+		ICMPType   uint8
+		ICMPCode   uint8
+		ICMPData   []byte
+		inPort     uint32
+		outPort    uint32
+	)
+
+	// Get ethernet data.
+	srcMAC = pktIn.Data.HWDst
+	dstMAC = pktIn.Data.HWSrc
+	IPProtocol = uint8(pktIn.Data.Ethertype)
+
+	switch pktIn.Data.Ethertype {
+	case protocol.IPv4_MSG:
+		// Get IP data.
+		ipv4In := pktIn.Data.Data.(*protocol.IPv4)
+		srcIP = ipv4In.NWDst.String()
+		dstIP = ipv4In.NWSrc.String()
+		switch ipv4In.Protocol {
+		case protocol.Type_TCP:
+			IPProtocol = protocol.Type_TCP
+
+			// Transfer Buffer to TCP.
+			tcpBytes, err := ipv4In.Data.(*util.Buffer).MarshalBinary()
+			if err != nil {
+				return err
+			}
+			tcpIn := new(protocol.TCP)
+			err = tcpIn.UnmarshalBinary(tcpBytes)
+			if err != nil {
+				return err
+			}
+
+			// Set TCP Reset packet data.
+			TCPSeqNum = 0
+			TCPAckNum = tcpIn.SeqNum + 1
+			TCPSrcPort = tcpIn.PortDst
+			TCPDstPort = tcpIn.PortSrc
+			TCPFlags = TCPRst + TCPAck
+		default:
+			// To follow the NSX-T implementation, use ICMP host administratively prohibited
+			// for ICMP, UDP, SCTP reject.
+			IPProtocol = protocol.Type_ICMP
+			// Set ICMP host administratively prohibited data.
+			ICMPType = ICMPDstUnreachableType
+			ICMPCode = ICMPDstHostAdminProhibitedCode
+			ipHdr, _ := ipv4In.MarshalBinary()
+			ICMPData = make([]byte, int(ICMPUnusedHdrLen+IPv4HdrLen+8))
+			// Put ICMP unused header in Data prop and set it to zero.
+			binary.BigEndian.PutUint32(ICMPData[:ICMPUnusedHdrLen], 0)
+			copy(ICMPData[ICMPUnusedHdrLen:], ipHdr[:IPv4HdrLen+8])
+		}
+	case protocol.IPv6_MSG:
+		ipv6In := pktIn.Data.Data.(*protocol.IPv6)
+		srcIP = ipv6In.NWDst.String()
+		dstIP = ipv6In.NWSrc.String()
+		switch ipv6In.NextHeader {
+		case protocol.Type_TCP:
+			IPProtocol = protocol.Type_TCP
+
+			// Transfer Buffer to TCP
+			tcpBytes, err := ipv6In.Data.(*util.Buffer).MarshalBinary()
+			if err != nil {
+				return err
+			}
+			tcpIn := new(protocol.TCP)
+			err = tcpIn.UnmarshalBinary(tcpBytes)
+			if err != nil {
+				return err
+			}
+			// Set TCP Reset packet data.
+			TCPSeqNum = 0
+			TCPAckNum = tcpIn.SeqNum + 1
+			TCPSrcPort = tcpIn.PortDst
+			TCPDstPort = tcpIn.PortSrc
+			TCPFlags = TCPRst + TCPAck
```

Comment:
These seems to be identical to the ipv4 case. We can write a function that takes an interface, which will work on both ipv4 and ipv6?

Reply:
Good point. Changed it.